### PR TITLE
feat(cloud-sources): discovered cluster matching

### PR DESCRIPTION
**Pull request by dhaus67**
## Description

This PR adds the matching logic to match discovered clusters with existing secured clusters.

This will ensure that discovered clusters will have their status (i.e. whether they are already secured or not) populated based on existing secured clusters known to Central.

Matching currently works as follows:
- In case secured clusters do not have any `ProviderMetadata` associated with them, they will not be considered for matching.
- In case secured clusters have partial `ProviderMetadata` (this can mean either that the `ClusterMetadata` type is unspecified _or_ the name / ID of the cluster is empty), the provider type will be marked as `Unspecified` during matching. The reason we decided to go with this is to avoid having false-positives (i.e. mark a cluster Unsecured even though it _may_ be secured).
- In case the secured cluster has all we need for matching in its `ProviderMetadata`, we create an index out of the cluster ID, name, and type. This will then be used to match against discovered clusters from cloud sources.

Each cloud source will then be associated with its type accordingly.

## Checklist
- [ ] Investigated and inspected CI test results
- [ ] Unit test and regression tests added
- [ ] Evaluated and added CHANGELOG entry if required
- [ ] Determined and documented upgrade steps
- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

If any of these don't apply, please comment below.

### Here I tell how I validated my change

- see added unit tests.

1. Create a valid Cloud source integration (e.g. Paladin one).

2. Observe within the logs after creation that clusters have been discovered
```log
Got the following discovered clusters: ...
```

3. Check the `v1/discovered-clusters` endpoint and see the clusters correctly matched:
```json
{
  "clusters": [
    {
      "id": "d24abbe8-e52d-53f4-bdd2-178efab6c3a5",
      "metadata": {
...
      },
      "status": "STATUS_UNSECURED",
      "source": {
        "id": "0c31f1f5-6082-4bed-a158-dba032a9d969"
      }
    },
    {
      "id": "bd032f2e-42b8-52cd-b1ba-203460b0431b",
      "metadata": {
...
      },
      "status": "STATUS_UNSECURED",
      "source": {
        "id": "0c31f1f5-6082-4bed-a158-dba032a9d969"
      }
    },
    {
      "id": "6b70865b-2eb7-5ae4-bad6-882fab89cd10",
      "metadata": {
...
      },
      "status": "STATUS_UNSECURED",
      "source": {
        "id": "0c31f1f5-6082-4bed-a158-dba032a9d969"
      }
    },
    {
      "id": "1da3a02c-54b7-5202-a89b-e927ebb7e991",
      "metadata": {
...
      },
      "status": "STATUS_UNSECURED",
      "source": {
        "id": "0c31f1f5-6082-4bed-a158-dba032a9d969"
      }
    },
    {
      "id": "8f451291-6db5-5a20-bdb9-abec4b76b859",
      "metadata": {
...
      },
      "status": "STATUS_UNSECURED",
      "source": {
        "id": "0c31f1f5-6082-4bed-a158-dba032a9d969"
      }
    }
  ]
}
```

### Reminder for reviewers

In addition to reviewing code here, reviewers **must** also review testing and request further testing in case the
performed one does not seem sufficient. As a reviewer, you must not approve the change until you understand the
performed testing and you are satisfied with it.

### CODE REVIEW

**Comment by dhaus67** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/9813?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

Current dependencies on/for this PR:
* `master`
  * **PR #9736** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/9736?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #9737** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/9737?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
      * **PR #9778** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/9778?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
        * **PR #9813** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/9813?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈
      * **PR #9769** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/9769?utm_source=stack-comment-icon" target="_blank">Graphite</a> 

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/stackrox/stackrox/9813?utm_source=stack-comment).

**Comment by stehessel** [comment]
> In case secured clusters have partial ProviderMetadata (this can mean either that the ClusterMetadata type is unspecified or the name / ID of the cluster is empty), the provider type will be marked as Unspecified during matching. The reason we decided to go with this is to avoid having false-positives (i.e. mark a cluster Unsecured even though it may be secured).

I wonder if we should base the matching on the cluster type instead of the provider type. Reason being that once we add OCM and OpenShift matching, the provider can be ambiguous.

File: central/cloudsources/manager/manager.go [hunk 1]
```diff
@@ -138,18 +150,64 @@ func (m *managerImpl) getDiscoveredClustersFromCloudSources() []*discoveredclust
...
+	debugPrintDiscoveredClusters(clusters)
 
 	if err := m.discoveredClustersDataStore.UpsertDiscoveredClusters(discoveredClusterCtx, clusters...); err != nil {
 		log.Errorw("Received errors during upserting discovered clusters.", logging.Err(err))
 	}
 }
 
+func (m *managerImpl) matchDiscoveredClusters(clusters []*discoveredclusters.DiscoveredCluster) {
+	// A list of hashes of currently secured cluster. A secured cluster hash consist of:
+	//  The secured cluster ID, name, and type.
```
Comment (stehessel) [inline central/cloudsources/manager/manager.go:169]: Technically cluster ID and type are sufficient, right? Why add name as well?

File: central/cloudsources/manager/manager.go [hunk 2]
```diff
@@ -138,18 +150,64 @@ func (m *managerImpl) getDiscoveredClustersFromCloudSources() []*discoveredclust
...
+	securedClusters := set.NewStringSet()
+
+	unspecifiedClusterTypes := set.NewStringSet()
+
+	if err := m.clusterDataStore.WalkClusters(clustersCtx, func(obj *storage.Cluster) error {
+		providerMetadata := obj.GetStatus().GetProviderMetadata()
+
+		// Explicitly ignore clusters which do not have provider metadata associated with them.
+		if providerMetadata == nil {
```
Comment (stehessel) [inline central/cloudsources/manager/manager.go:178]: This will be a source of false positives because of old sensors. But I guess we can't do anything about it otherwise than tell customers to upgrade their sensors to get correct matching.

File: central/cloudsources/manager/manager.go [hunk 3]
```diff
@@ -138,18 +150,64 @@ func (m *managerImpl) getDiscoveredClustersFromCloudSources() []*discoveredclust
...
+			return nil
+		}
+
+		// In case we have partial provider metadata (e.g. we miss the cluster name for EKS clusters, or the
+		// cluster type is unspecified), then mark the provider type as Unspecified.
+		// This means that we will assign each discovered cluster that cannot be safely matched to a secured cluster
+		// as Unspecified instead of Unsecured. This is to avoid false-positives.
+		if stringutils.AtLeastOneEmpty(providerMetadata.GetCluster().GetId(), providerMetadata.GetCluster().GetName()) {
```
Comment (stehessel) [inline central/cloudsources/manager/manager.go:186]: Can extract `providerMetadata.GetCluster()` into variable here.

File: central/cloudsources/manager/manager.go [hunk 4]
```diff
@@ -160,3 +218,21 @@ func createClients(cloudSources []*storage.CloudSource) []cloudsources.Client {
 	}
 	return clients
 }
+
+type clusterIndex = string
+
+func clusterIndexForClusterMetadata(obj *storage.ClusterMetadata) clusterIndex {
+	return obj.GetId() + obj.GetName() + obj.GetType().String()
```
Comment (stehessel) [inline central/cloudsources/manager/manager.go:225]: I wonder how long those names can get. Maybe hashing is safer?

File: central/cloudsources/manager/manager_test.go [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+	for _, cluster := range m.clusters {
+		_ = fn(cluster)
+	}
+	return nil
+}
+
+func TestMatchDiscoveredClusters(t *testing.T) {
+	clusterStore := &mockClusterStore{clusters: []*storage.Cluster{
+		createCluster(
+			"MC_testing_test-cluster-1_eastus",
```
Comment (stehessel) [inline central/cloudsources/manager/manager_test.go:30]: missing subscription id

File: central/cloudsources/manager/manager_test.go [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+			"test-cluster-1",
+			storage.ClusterMetadata_AKS,
+		),
+		createCluster("something-else", "something-else", storage.ClusterMetadata_ROSA),
```
Comment (stehessel) [inline central/cloudsources/manager/manager_test.go:34]: I would suggest to use the real schema for `id`. I.e. uuid4 for openshift, arn for EKS.

File: central/cloudsources/manager/manager_test.go [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+			Type:         storage.ClusterMetadata_EKS,
+			ProviderType: storage.DiscoveredCluster_Metadata_PROVIDER_TYPE_AWS,
+		},
+		{
+			ID:           "5553424234234_MC_testing_unsecured_eastus",
+			Name:         "unsecured",
+			Type:         storage.ClusterMetadata_AKS,
+			ProviderType: storage.DiscoveredCluster_Metadata_PROVIDER_TYPE_AZURE,
+		},
+		{
```
Comment (stehessel) [inline central/cloudsources/manager/manager_test.go:88]: Maybe I'm missing something, but shouldn't this cluster be `secured` since it's added above to the clusters?

Comment (stehessel) [reply central/cloudsources/manager/manager_test.go:88]: Ah I see. But should we require name, given that we don't even need it for matching? I think requiring ID and type is enough.

File: central/cloudsources/manager/manager.go [hunk 8]
```diff
@@ -138,18 +150,65 @@ func (m *managerImpl) getDiscoveredClustersFromCloudSources() []*discoveredclust
...
+		if providerMetadata == nil {
+			return nil
+		}
+
+		// In case we have partial provider metadata (e.g. we miss the cluster name for EKS clusters, or the
+		// cluster type is unspecified), then mark the provider type as Unspecified.
+		// This means that we will assign each discovered cluster that cannot be safely matched to a secured cluster
+		// as Unspecified instead of Unsecured. This is to avoid false-positives.
+		clusterMetadata := providerMetadata.GetCluster()
+		if stringutils.AtLeastOneEmpty(clusterMetadata.GetId(), clusterMetadata.GetName()) {
```
Comment (stehessel) [inline central/cloudsources/manager/manager.go:187]: I'm not sure this does the correct thing now. Afaik we want the following behavior (for AWS specifically, but also generally):

Secured `/v1/clusters`:
* 1 EKS cluster 'A' with metadata ID and type correctly set.
* 1 EKS cluster 'B' with without metadata ID and type. This happens when instance tags are disabled, and results in `providerMetadata.Provider` being set, but `providerMetadata.ClusterMetadata` being nil.
* 1 OpenShift cluster 'C' being hosted on AWS with correct metadata.

Discovered clusters:
* 1 EKS cluster that matches 'A'. This should show as secured.
* 1 EKS cluster without match. This should show as unspecified.
* 1 OpenShift cluster on AWS (from OCM coming soon™) matching 'C'. This should show as secured.
* 1 OpenShift cluster on AWS without match. This should show as unsecured.

So in summary, EKS clusters in `/v1/clusters` may show up without Metadata type. So your idea with matching on provider level may be more correct after all.

**Review by stehessel — APPROVED** [review]
